### PR TITLE
Fix  : Unable to type next paragraph in message

### PR DESCRIPTION
**Pull request by ahmdshrif**
### Fixed Issues

$ https://github.com/Expensify/App/issues/6767

### Tests/QA Steps

1- Go to app
2- Open any chat
3- Type 1 line & hit return to add a paragraph in same message

- [x] verify that enter make a new line in the mobile web and touchable device. 
- [x] very that enter send massage for the desktop and desktop web. 
- [x] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web

https://user-images.githubusercontent.com/21364901/150486159-76a1a06b-14bc-4dfa-8a0e-eb9770095087.mov

#### Mobile Web

https://user-images.githubusercontent.com/21364901/150493414-bca70d88-3123-4a16-b28e-bda938084532.mp4

#### Desktop

https://user-images.githubusercontent.com/21364901/150486192-09dfa255-60cd-4560-9e31-a5eba2b674c4.mov

#### iOS

https://user-images.githubusercontent.com/21364901/150501756-ff4ac205-3df1-441b-9521-0d0324b03763.mov

#### Android

https://user-images.githubusercontent.com/21364901/150488677-eed72684-391a-47eb-b8de-e312287e0318.mov

### CODE REVIEW

File: src/pages/home/report/ReportActionCompose.js [hunk 1]
```diff
@@ -353,13 +355,27 @@ class ReportActionCompose extends React.Component {
         }
     }
 
+    /**
+     * Allows us to identify whether Virtual keyboard open or not
+     *
+     * @returns {boolean}
```
Comment (parasharrajat) [inline src/pages/home/report/ReportActionCompose.js:361]:
```suggestion
     * Whether virtual keyboard is open or not
     * @returns {Boolean}
```

File: src/pages/home/report/ReportActionCompose.js [hunk 2]
```diff
@@ -146,6 +147,7 @@ class ReportActionCompose extends React.Component {
         this.setTextInputRef = this.setTextInputRef.bind(this);
         this.getInputPlaceholder = this.getInputPlaceholder.bind(this);
         this.setPreferredSkinTone = this.setPreferredSkinTone.bind(this);
+        this.isVirtualKeyboardOpen = this.isVirtualKeyboardOpen.bind(this);
```
Comment (parasharrajat) [inline src/pages/home/report/ReportActionCompose.js:150]:
```suggestion
```

File: src/pages/home/report/ReportActionCompose.js [hunk 3]
```diff
@@ -353,13 +355,27 @@ class ReportActionCompose extends React.Component {
...
+    /**
+     * Allows us to identify whether Virtual keyboard open or not
+     *
+     * @returns {boolean}
+     */
+    isVirtualKeyboardOpen() {
+        // check if platform is web and supports virtualkeyboard
+        if (navigator && 'virtualkeyboard' in navigator) {
+            const keyboardPosition = navigator.virtualKeyboard.boundingRect.y;
+            return keyboardPosition > 0;
```
Comment (parasharrajat) [inline src/pages/home/report/ReportActionCompose.js:367]:
```suggestion
            return navigator.virtualKeyboard.boundingRect.y > 0;
```

File: src/pages/home/report/ReportActionCompose.js [hunk 4]
```diff
@@ -353,13 +355,27 @@ class ReportActionCompose extends React.Component {
...
     }
 
+    /**
+     * Allows us to identify whether Virtual keyboard open or not
+     *
+     * @returns {boolean}
+     */
+    isVirtualKeyboardOpen() {
+        // check if platform is web and supports virtualkeyboard
+        if (navigator && 'virtualkeyboard' in navigator) {
```
Comment (parasharrajat) [inline src/pages/home/report/ReportActionCompose.js:365]:
```suggestion
        if (navigator && 'virtualKeyboard' in navigator) {
```

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/pages/home/report/ReportActionCompose.js [hunk 5]
```diff
@@ -353,13 +354,25 @@ class ReportActionCompose extends React.Component {
         }
     }
 
+    /**
+     * Whether virtual keyboard is open or not
+     * @returns {Boolean}
+     */
+    isVirtualKeyboardOpen() {
+        // check if platform is web and supports virtualkeyboard
```
Comment (parasharrajat) [inline src/pages/home/report/ReportActionCompose.js:362]:
```suggestion
        // Whether the virtualKeyboard API is supported on the platform.
```

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/pages/home/report/ReportActionCompose.js (hunk 5 above)
Comment (ahmdshrif) [reply src/pages/home/report/ReportActionCompose.js]: ok done

**Review by parasharrajat — DISMISSED** [review]
LGTM. 

cc: @roryabraham 
:ribbon: :eyes: :ribbon:  C+ reviewed

File: src/pages/home/report/ReportActionCompose.js [hunk 6]
```diff
@@ -353,13 +354,25 @@ class ReportActionCompose extends React.Component {
         }
     }
 
+    /**
+     * Whether virtual keyboard is open or not
+     * @returns {Boolean}
+     */
+    isVirtualKeyboardOpen() {
```
Comment (roryabraham) [inline src/pages/home/report/ReportActionCompose.js:361]: Can you please move this to it's own `VirtualKeyboard` library to make it more discoverable?

Comment (roryabraham) [reply src/pages/home/report/ReportActionCompose.js:361]: You should have it return `true` or `false` if the VirtualKeyboard API is available, or null if it's not. Do not fall back on `canUseTouchScreen` within the library.

Like this:
#### src/libs/virtualKeyboard.js
```js
/**
 * Is the virtual keyboard open?
 *
 * @returns {Boolean|null} – null if the VirtualKeyboard API is unavailable
 */
function isOpen() {
    if (!_.has(navigator, 'virtualKeyboard')) {
        return null;
    }
    return navigator.virtualKeyboard.boundingRect.y > 0;
}

export {
    isOpen,
};
```

#### src/pages/home/report/ReportActionCompose.js
```js
...
...
import * as VirtualKeyboard from '../../../libs/virtualKeyboard';
...
...

/**
 * @returns {Boolean}
 */
shouldAssumeVirtualKeyboardIsOpen() {
    const isOpen = VirtualKeyboard.isOpen();
    return _.isNull(isOpen) ? canUseTouchScreen() : isOpen;
}

/**
 * Listens for keyboard shortcuts and applies the action
 *
 * @param {Object} e
 */
triggerHotkeyActions(e) {
    if (!e || this.shouldAssumeVirtualKeyboardIsOpen()) {
        return;
    }
    ...
    ...
}
```

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: src/pages/home/report/ReportActionCompose.js (hunk 6 above)
Comment (parasharrajat) [reply src/pages/home/report/ReportActionCompose.js:361]: +1.

Comment (ahmdshrif) [reply src/pages/home/report/ReportActionCompose.js]: ok done

**Review by parasharrajat — DISMISSED** [review]

File: src/pages/home/report/ReportActionCompose.js [hunk 7]
```diff
@@ -353,13 +355,21 @@ class ReportActionCompose extends React.Component {
         }
     }
 
+    /**
+     * @returns {boolean}
```
Comment (roryabraham) [inline src/pages/home/report/ReportActionCompose.js:359]:
```suggestion
     * @returns {Boolean}
```

File: src/pages/home/report/ReportActionCompose.js [hunk 8]
```diff
@@ -353,13 +355,21 @@ class ReportActionCompose extends React.Component {
...
+     */
+    shouldAssumeVirtualKeyboardIsOpen() {
+        const isOpen = VirtualKeyboard.isOpen();
```
Comment (roryabraham) [inline src/pages/home/report/ReportActionCompose.js:362]: Let's add a comment here:

```
/* 
 * As of January 2022, the VirtualKeyboard web API is not available in all browsers yet
 * If it is unavailable, we default to assuming that the virtual keyboard is open on touch-enabled devices.
 * See https://github.com/Expensify/App/issues/6767 for additional context.
 */ 
```

File: src/libs/virtualKeyboard.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import _ from 'underscore';
```
Comment (roryabraham) [inline src/libs/virtualKeyboard.js:1]: I thought about this some more, and I think it's a bad idea to create this library with only the web implementation when we can easily add a native implementation. That way, nobody will be surprised if they later attempt to use this library in another context and find that it always returns `null` in the native platforms. To implement this, simply do the following:

1. Move this file to `src/libs/VirtualKeyboard/index.js`
1. Create a new file called `src/libs/VirtualKeyboard/index.native.js`, and give it the following content:

```javascript
let isVirtualKeyboardOpen = false;

Keyboard.addListener(
    'keyboardDidShow',
    () => {
      isVirtualKeyboardOpen = true;
    },
);

Keyboard.addListener(
    'keyboardDidHide',
    () => {
        isVirtualKeyboardOpen = false;
    },
);

/**
 * Is the virtual keyboard open?
 *
 * Note – the web equivalent of this function may return null.
 *
 * @returns {Boolean}
 */
function isOpen() {
    return isVirtualKeyboardOpen;
};

export {
    // eslint-disable-next-line import/prefer-default-export
    isOpen,
};
```

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: src/libs/virtualKeyboard.js (hunk 9 above)
Comment (ahmdshrif) [reply src/libs/virtualKeyboard.js:1]: yes, it will be helpful in the case of a tablet with a physical keyboard. using a native app.

Comment (ahmdshrif) [reply src/libs/virtualKeyboard.js]: it'd done now @roryabraham

**Review by roryabraham — APPROVED** [review]
Looks good and tests well

**Comment by roryabraham** [comment]
@ahmdshrif Thanks for making those few updates I requested 👍

**Comment by ahmdshrif** [comment]
thanks @roryabraham

**Comment by Tushu17** [comment]
@MitchExpensify, can you hire me for reporting bonus. I have submitted the proposal but it's not showing up for some reason

**Pull request merged**